src/chainstate/stacks/boot/pox.clar

```diff
@@ -0,0 +1,754 @@
+(define-constant ERR-STACKING-INSUFFICIENT-FUNDS 1)
+(define-constant ERR-STACKING-INVALID-LOCK-PERIOD 2)
+(define-constant ERR-STACKING-ALREADY-STACKED 3)
+(define-constant ERR-STACKING-NO-SUCH-PRINCIPAL 4)
+(define-constant ERR-STACKING-EXPIRED 5)
+(define-constant ERR-STACKING-STX-LOCKED 6)
+(define-constant ERR-STACKING-NO-SUCH-DELEGATE 7)
+(define-constant ERR-STACKING-BAD-DELEGATE 8)
+(define-constant ERR-STACKING-PERMISSION-DENIED 9)
+(define-constant ERR-STACKING-INVALID-DELEGATE-TENURE 10)
+(define-constant ERR-STACKING-THRESHOLD-NOT-MET 11)
+(define-constant ERR-STACKING-POX-ADDRESS-IN-USE 12)
+(define-constant ERR-STACKING-INVALID-POX-ADDRESS 13)
+(define-constant ERR-STACKING-ALREADY-DELEGATED 14)
+(define-constant ERR-STACKING-DELEGATE-ALREADY-REGISTERED 15)
+(define-constant ERR-STACKING-DELEGATE-EXPIRED 16)
+(define-constant ERR-STACKING-ALREADY-REJECTED 17)
+(define-constant ERR-STACKING-INVALID-AMOUNT 18)
+
+;; Min/max number of reward cycles uSTX can be locked for
+(define-constant MIN-POX-REWARD-CYCLES u1)
+(define-constant MAX-POX-REWARD-CYCLES u12)
+
+;; Default length of the PoX registration window, in burnchain blocks.
+(define-constant REGISTRATION-WINDOW-LENGTH u250)
+
+;; Default length of the PoX reward cycle, in burnchain blocks.
+(define-constant REWARD-CYCLE-LENGTH u1000)
+
+;; Valid values for burnchain address versions.
+;; These correspond to address hash modes in Stacks 2.0.
+(define-constant ADDRESS-VERSION-P2PKH u0)
+(define-constant ADDRESS-VERSION-P2SH u1)
+(define-constant ADDRESS-VERSION-P2WPKH u2)
+(define-constant ADDRESS-VERSION-P2WSH u3)
+
+;; Stacking thresholds
+(define-constant STACKING-THRESHOLD-25 u20000)
+(define-constant STACKING-THRESHOLD-100 u5000)
+
+;; PoX disabling threshold (a percent)
+(define-constant POX-REJECTION-FRACTION u25)
+
+;; Data vars that store a copy of the burnchain configuration.
+;; Implemented as data-vars, so that different configurations can be
+;; used in e.g. test harnesses.
+(define-data-var pox-registration-window-length uint REGISTRATION-WINDOW-LENGTH)
+(define-data-var pox-reward-cycle-length uint REWARD-CYCLE-LENGTH)
+(define-data-var pox-rejection-fraction uint POX-REJECTION-FRACTION)
+(define-data-var first-burnchain-block-height uint u0)
+(define-data-var configured bool false)
+
+;; This function can only be called once, when it boots up
+(define-public (set-burnchain-parameters (first-burn-height uint) (pox-reg-window-len uint) (pox-reward-cycle-len uint) (pox-rejection-frac uint))
+    (if (and is-in-regtest (not (var-get configured)))
+        (begin
+            (var-set first-burnchain-block-height first-burn-height)
+            (var-set pox-registration-window-length pox-reg-window-len)
+            (var-set pox-reward-cycle-length pox-reward-cycle-len)
+            (var-set pox-rejection-fraction pox-rejection-frac)
+            (var-set configured true)
+            (ok true)
+        )
+        (ok false)
```

Comment:
Should this function returns an `(err ...)`, instead of `(ok false)`?

Reply:
I'm ambivalent -- the method's a no-op outside of the test framework.  I can make it `(err ..)`.

Reply:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/1850/commits/1f83846f4edc21eb24cad85d9bfae1f422fe33f2